### PR TITLE
[quant][graphmode] linear_relu

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36994 [quant][graphmode] Test refactor -- extract graph checker
* **#36889 [quant][graphmode] linear_relu**

Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

### CODE REVIEW

File: torch/quantization/qconfig.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ class QConfig(namedtuple('QConfig', ['activation', 'weight'])):
     Observer classes have usually reasonable default arguments, but they can be overwritten with `with_args`
     method (that behaves like functools.partial):
 
-      my_qconfig = QConfig(activation=MinMaxObserver.with_args(dtype=torch.qint8), 
+      my_qconfig = QConfig(activation=MinMaxObserver.with_args(dtype=torch.qint8),
```
Comment (z-a-f) [inline torch/quantization/qconfig.py:21]: Sorry, this one happened because my editor removes all the end lines, and this file was open :/

File: test/quantization/test_quantize_script.py [hunk 2]
```diff
@@ -1169,12 +1169,58 @@ def _test_op_impl(self, module, data, quantized_op):
         model = quantize_script(model, qconfig_dict, _test_only_eval_fn, [data], inplace=False)
         FileCheck().check(quantized_op) \
                    .run(model.graph)
-
         # make sure it runs
         model(data[0][0])
 
         return model
 
+    @unittest.skipUnless(
```
Comment (raghuramank100) [inline test/quantization/test_quantize_script.py:1177]: Do we need to skip QNNPACK?

File: test/quantization/test_quantize_script.py [hunk 3]
```diff
@@ -1169,12 +1169,58 @@ def _test_op_impl(self, module, data, quantized_op):
...
+
+    @unittest.skipUnless('fbgemm' in torch.backends.quantized.supported_engines,
+                         " Quantized operations require FBGEMM. FBGEMM is only optimized for CPUs"
+                         " with instruction set support avx2 or newer.")
+    def test_quantized_linear_relu(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super(M, self).__init__()
+                self.linear = torch.nn.Linear(30, 4).float()
+                self.relu = torch.nn.ReLU()
```
Comment (raghuramank100) [inline test/quantization/test_quantize_script.py:1209]: Do we need to test functional relu as well?
For completeness, should we also test functional linear + functional relu?

File: test/quantization/test_quantize_script.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1177]: not sure if qnnpack works, my backward compatibility PR added qnnpack and the test failed

Comment (z-a-f) [reply test/quantization/test_quantize_script.py:1178]: When I enable the qnnpack locally, the tests pass

Comment (z-a-f) [reply test/quantization/test_quantize_script.py:1178]: I take it back -- the tests fail. Specifically line 950 in `observer.py` throws `RuntimeError: RuntimeError: Trying to create tensor with negative dimension -4398046511104`. I think that might be related to #36888

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -667,6 +667,20 @@ class InsertObserversHelper {
   // These are the IR patterns we match to skip inserting observers.
   // They are compiled once on construction and used repeatedly within
   // the pass.
+  const PatternInfo linear_functional_relu = PatternInfo::parse_from_str(R"(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:670]: maybe add the fusion for F.linear + F.relu and F.linear + self.relu(module) here?
We may need these two patterns for F.conv as well, but it can be in a different PR.

Comment (z-a-f) [reply torch/csrc/jit/passes/quantization.cpp:670]: OK, I will look into this

File: test/quantization/test_quantize_script.py [hunk 5]
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
...
+    def test_quantized_functional_linear(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super(M, self).__init__()
+                self.linear = F.linear
+                self.w = torch.randn(4, 30)
+                self.b = torch.randn(4)
+
+            def forward(self, x):
+                return self.linear(x, self.w, self.b)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1215]: you can just call F.linear here

File: test/quantization/test_quantize_script.py [hunk 6]
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
...
+                             .check_not("quantized::relu(")
+        checker.run(model.graph)
+        checker.run(model_functional.graph)
+
+    @unittest.skipUnless(
+        'fbgemm' in torch.backends.quantized.supported_engines,
+        " Quantized operations require FBGEMM. FBGEMM is only optimized for CPUs"
+        " with instruction set support avx2 or newer.",
+    )
+    def test_quantized_functional_linear_relu(self):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1263]: actually, can you merge this test with the previous one?

File: torch/csrc/jit/passes/quantization.cpp [hunk 7]
```diff
@@ -766,6 +801,10 @@ graph(%self, %a, %b, %inplace):
 
   const std::vector<std::reference_wrapper<const PatternInfo>> delay_patterns =
       {
+          nn_linear_nn_relu,
+          nn_linear_f_relu,
+          f_linear_nn_relu,
+          f_linear_f_relu,
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:807]: this looks like a great naming convention, we should probably follow the same for other patterns in the list.

File: test/quantization/test_quantize_script.py (hunk 6 above)
Comment (z-a-f) [reply test/quantization/test_quantize_script.py:1263]: I am not sure I can -- I tried doing it, but that creates if statements in the forward, which fails to match the patterns :'(

File: torch/csrc/jit/passes/quantization.cpp (hunk 7 above)
Comment (z-a-f) [reply torch/csrc/jit/passes/quantization.cpp:807]: Will do when adding it to conv

**Review by jerryzh168 — APPROVED** [review]
It's fine to keep the tests separate if it's hard to merge them

File: test/quantization/test_quantize_script.py [hunk 8]
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
...
+        model = self._test_op_impl(M(functional=False), data,
+                                   "quantized::linear_relu")
+        model_functional = self._test_op_impl(M(functional=True), data,
+                                              "quantized::linear_relu")
+        checker = FileCheck().check_not("aten::linear") \
+                             .check_not("aten::relu") \
+                             .check_not("quantized::linear(") \
+                             .check_not("quantized::relu(")
+        checker.run(model.graph)
+        checker.run(model_functional.graph)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1256]: probably better to loop over True and False here and put the call to _test_op_impl and checker.run in the loop

File: test/quantization/test_quantize_script.py [hunk 9]
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
...
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1289]: same here

File: torch/csrc/jit/passes/quantization.cpp [hunk 10]
```diff
@@ -665,9 +665,44 @@ class InsertObserversHelper {
 
   // Is dynamic quantization enabled for the observer pass.
   bool is_dynamic = false;
+
   // These are the IR patterns we match to skip inserting observers.
   // They are compiled once on construction and used repeatedly within
   // the pass.
+
+  // nn.Linear + nn.ReLU
+  const PatternInfo nn_linear_nn_relu = PatternInfo::parse_from_str(R"(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization.cpp:674]: btw these patterns will need to be updated, please see the code in master to see how to do the update

**Comment by z-a-f** [comment]
Replaced by #40021
